experiments/04B.md

```diff
@@ -0,0 +1,234 @@
+| 2    | 61  | 23.98 | 44.9	  | 20.92	 | فادع لنا ربك يخرج لنا مما تنبت الأرض من بقلها وقفائها وفومها وعدسها وبصلها	| وفومها وعدسها وبصلها |	-11|	        0 |	0  |
+| 2    | 61  | 44.9  | 53.96  | 9.06	 | قال أتستبدلون الذي هو أدنى بالذي هو أدنى |	قال أتستبدلون الذي هو أدنى |	-3|	0|	0 |
+| 2    | 61  | 53.98 | 64.32  | 10.34	 | إهبطوا مصرا فإن لكم ما سألتم	بالذي هو خير | اهبطوا مصرا فإن لكم ما سألتم |	3|	0|	0 |
+| 2    | 61  | 64.32 | 77.88  | 13.56	 | وضربت عليهم الذلة والمسكنة وباءوا بغضب من الله |	وضربت عليهم الذلة والمسكنة وباءوا بغضب من الله |	0|	0|	0 |
+| 2    | 61  | 77.88 | 83.96  | 6.08     | ذلك بأنهم كانوا مصرين |	ذلك بأنهم كانوا يكفرون |	0|	0|	0 |
+| 2    | 61  | 83.98 | 94.42  | 10.44	 | يكفرون بآيات الله ويقتلون النبيين بغير الحق |	بآيات الله ويقتلون النبيين بغير الحق |	-1|	0|	0 |
+| 2    | 61  | 94.42 | 101.64 | 7.22	 | ذلك بما عصوا وكانوا يعتدون |	ذلك بما عصوا وكانوا يعتدون |	0|	0|	0|
+
+
+The shiekh said the part: "فادع لنا ربك يخرج لنا مما تنبت الأرض من بقلها وقثائها" twice. The algorithm didn't work well with this aya. The reference text is not correct. The algorithm should be modified to handle this case. (This case still exists in the current version of the algorithm)
+
+
+## SubExperiment: Repeated parts in the reference text
+### Objective
+I found out that I have to see the repeated parts' cases in the reference text of the Holy Quran to cover those cases.
+
+### Code [[2]]
+```python
+repeated_substrings = {} # key:(sura, aya), value: list of repeated substrings
+phras_repeated = [0]*len(reference_dataframe)
+for i in trange(len(reference_dataframe)):
+    sura = reference_dataframe['sorah'][i]
+    aya = reference_dataframe['ayah'][i]
+    aya_text = " " + reference_dataframe['text'][i] + " "
+    aya_text_split = aya_text.split()
+    repeated_substrings[(sura, aya)] = {}
+    for window_size in range(1, (len(aya_text_split)//2)+1):
+        for start_idx in range(len(aya_text_split)-window_size):
+            sub = " ".join(aya_text_split[start_idx:start_idx+window_size])
+            num = aya_text.count(' '+sub+' ', len(' '.join(aya_text_split[:start_idx])))
+            if num > 1:
+                repeated_substrings[(sura, aya)][sub] = num
+                if window_size > 1:
+                    phras_repeated[i] += 1
+```
+
+### Results
+
+```py
+for i in range(0, 10):
+    n = len(diff_df[diff_df['phrase_repeated'] > i])
+    print(f"{n}\tayas that have {i+1}\tof repeated prases")
+```
+```
+277	ayas that have  1	 or more of repeated prases
+91	ayas that have  2	 or more of repeated prases
+72	ayas that have  3	 or more of repeated prases
+43	ayas that have  4	 or more of repeated prases
+29	ayas that have  5	 or more of repeated prases
+26	ayas that have  6	 or more of repeated prases
+12	ayas that have  7	 or more of repeated prases
+10	ayas that have  8	 or more of repeated prases
+10	ayas that have  9	 or more of repeated prases
+8	ayas that have  10   or more of repeated prases
+```

```

Comment:
This should be a table
|number of repeated segments|occurrences|
|---------------------------------------------|-----------------|
|>1|277|

Reply:
Ok, It was ment to memic a notebook cell and it output.